### ISSUE TITLE
[l10n] Merge German translations

### ISSUE THREAD

**Issue by fllp** [sailfishos-patches/patchmanager#27]
There are currently three project languages for German:  
`German (Germany)`, `German (Austria)` and `German (de)`.  
This is unnecessary and having only one tranlsation (`German (de)`) should be considered:
- for one, there is no such distinction in the official SFOS translations
- also, it is unclear which version will end up in Patchmanager:
  How does the App determine the locale variant to be used once installed?
- And: In the worst case, this 'split' triples efforts for updating or improving translations

**Comment by CODeRUS** [comment]
It's up to users to request languages for translation and translate it. 
system language is used. My best guess it's a de_de one.

**Comment by nephros** [comment]
#52 should resolve this. Closing.

**Comment by Olf0** [comment]
Well, there are more to handle: https://github.com/sailfishos-patches/patchmanager/pull/46#issuecomment-933027718

**Comment by b100dian** [comment]
I've updated transifex with removal of `ca`, `de_AT`, `de_DE`, `en_FI`, `it`, `ja`.
I don't know yet how to handle e.g. moving `fr_FR` to `fr`

**Comment by nephros** [comment]
Let me tack on another question about the workflow:

As a dev who changes stuff in the source files, when do I run lupdate, and do I run it as:
`lupdate -pro patchmanger.pro -ts translations/settings-patchmanager.ts` for just the source file   
or  
`lupdate -pro patchmanger.pro` which will change all the -LANG.ts files as well?

And: do I include the changes in every commit, as a separate commit per-PR, or not at all? Because I imagine updating it often may cause a lot of unnecessary merge conflicts in the ts files while PRs or branches are worked on. (And they clutter the files-changed/review tab).

Perhaps the best strategy would actually to *only* update ts files against master, and only after a merge to master as a separate commit (or even automated somehow).

**Comment by b100dian** [comment]
I've continued deleting `fr_FR` and `pt_BR`, adding `fr` and `pt` instead (+Invited collaborators for the latter).
The `nl` vs `nl_BE` seem to differ ("Pleister" vs "Plakker: all over, plus others) - ~delaying decision a bit~ - actually removed as there were no members in nl_BE.
Also `fi` and `fi_FI` differ, but the latter does not have members, and its less translated, so gone it is (I have backups).
`sl` vs `sl_SL` was simple, they were the same, same member, and differing only in 1 string missing in sl_SL.
@Olf0 you accidentally `zh_CN`, shouldn't that be `zh`?

Also I've uploaded the `de` translation from the repo (since it didn't show up after integrating the strings - so it seems this is one-way, transifex=>github for non source-language, and github=>transifex for the source-language)

So this leaves us with the next steps to make the same clean-up in the git repo (`src/qml/settings-patchmanager.ts`  but also `translations/settings-patchmanager-en.ts` which is not used?)

---

@nephros 
>     * When do I run lupdate, and do I run it as:
>       `lupdate -pro patchmanger.pro -ts translations/settings-patchmanager.ts` for just the source file
>       or
>       `lupdate -pro patchmanger.pro` which will change all the -LANG.ts files as well?

The first one. I think it was not clear to me either on the first update. 
`lupdate -pro patchmanger.pro -ts translations/settings-patchmanager.ts`

>     * Do I include the changes in every commit, as a separate commit per-PR, or not at all?
>       Because I imagine updating often may cause a lot of unnecessary merge conflicts in the ts files while PRs or branches are worked on. (And they clutter the files-changed/review tab).

I'd say (from now on) - "none at all" - and we make a separate PR when things stabilize towards a release.

**Comment by Olf0** [comment]
> I've continued deleting `fr_FR` and `pt_BR`, adding `fr` and `pt` instead (+Invited collaborators for the latter). 

Thanks, especially for having the idea to invite the extant collaborators.

> The `nl` vs `nl_BE` seem to differ ("Pleister" vs "Plakker: all over, plus others) - 

Yes, I know; but from my point of view the differences between Dutch (nl_NL) and Flemish (nl_BE) are negligible and all Flemish people are used to read Dutch.
See, e.g. https://www.youtube.com/watch?v=VrmdY5eQUBw (the first 40 seconds or 6 minutes or full video of ~ 18 minutes). 
Or (the same guy) even shorter here: First 20 seconds or whole 11 minutes.

> ~delaying decision a bit~ - actually removed as there were no members in nl_BE.

:+1:

> Also `fi` and `fi_FI` differ, but the latter does not have members, and its less translated, so gone it is (I have backups).

Well, we have git any way: We always can look back.

> `sl` vs `sl_SL` was simple, they were the same, same member, and differing only in 1 string missing in sl_SL.

:+1:

> @Olf0 you accidentally `zh_CN`, shouldn't that be `zh`?

Well, there is a verb missing, but you are right: I missed that!
Maybe my mind suppressed this one, because it is politically and technically the most loaded one: See the first three of the of the answers (not counting the indented comments) here https://stackoverflow.com/questions/4892372/language-codes-for-simplified-chinese-and-traditional-chinese
So zh_CN = (formally, correct) zh_HANS = "Simplified Chinese / Simplified Han language" as used in "the Peoples Republic of China (PRC)", which is synonym to "mainland China".
and zh_TW = (formally, correct) zh_HANT = "Traditional Chinese / Traditional Han language" as used in "Republic of China (ROC)", which is synonym to Taiwan.

It is the only one case of languages with a common prefix (zh) I am aware of, where linguists actually agree that they have developed into different languages over the past 120 years, but most significantly since 1949 (i.e., since the manifestation of the political split: See "the Kuomintang / Chiang Kai-shek / White terror versus the Chinese Communist Party (CCP) / Mao Zedong / Red terror" story, in a brief form here https://en.wikipedia.org/wiki/Taiwan#Republic_of_China_(1945%E2%80%931949) and in details in the Wikipedia pages linked there.

Back to your question, "*Shouldn't `zh_CN` be `zh`*": I still would answer that with a bold YES, because
* "Simplified Chinese" will be well understood by "Traditional Chinese" speakers, even though some will not like to read it (emotionally / politically).
* We have no translator for "Traditional Chinese".
*  "Traditional Chinese" will not necessarily / easily be understood by "Simplified Chinese" speakers.
* The "Simplified Chinese" speakers outnumber the "Traditional Chinese" speakers by a factor of ~ 50.
* We want to keep things simple for us.
* IMO Patchmanager should not care too much about politics.

> Also I've uploaded the `de` translation from the repo (since it didn't show up after integrating the strings - so it seems this is one-way, transifex=>github for non source-language, and github=>transifex for the source-language)

O.K.; I may take a look at it at Transifex (or @nephros?), and also the other changes to translations [I proposed years ago](https://github.com/sailfishos-patches/patchmanager/pull/46), when your clean-up work has settled a bit.

> So this leaves us with the next steps to make the same clean-up in the git repo (`src/qml/settings-patchmanager.ts` but also `translations/settings-patchmanager-en.ts` which is not used?)

Ack; yes, I think so.

Ah, you did all this work only at Transifex so far!?!
I have been searching for branches in this and your Github-repo and found nothing: I will look at it on Transifex this weekend.

**Comment by b100dian** [comment]
Really thorough answer, sorry for triggering you to invest in a such lengthy research - but very interesting in the end!

> Ah, you did all this work only at Transifex so far!?!

Yes, only on transifex. The idea is that we can touch up the files in the repo later: e.g. when `fr` comes in, we amend the transifex PR with deletion of `fr_FR`. (We can do it before of course, I just spent enough time yesterday only on transifex to do it at once)

**Comment by b100dian** [comment]
I got it

**Comment by b100dian** [comment]
For reference, I used this button 
![image](https://user-images.githubusercontent.com/71930/136636626-d4b04c7e-9459-425d-ac31-a32884ddd07a.png)
in the header of the bottom list. Project Maintainer should have the ability to invite users per https://docs.transifex.com/teams/understanding-user-roles/#teams-&-collaboration

**Comment by Olf0** [comment]
> Really thorough answer, sorry for triggering you to invest in a such lengthy research - but very interesting in the end!

Glad you liked it!
(There was no research of content, just of links.) 
 
> > Ah, you did all this work only at Transifex so far!?!
> 
> Yes, only on transifex.

https://www.transifex.com/coderus/patchmanager3/languages/

> The idea is that we can touch up the files in the repo later: e.g. when `fr` comes in, we amend the transifex PR with deletion of `fr_FR`. (We can do it before of course, I just spent enough time yesterday only on transifex to do it at once)



> For reference, I used this button ![image](https://user-images.githubusercontent.com/71930/136636626-d4b04c7e-9459-425d-ac31-a32884ddd07a.png) in the header of the bottom list.

AFAICS one needs to be Team Manager for that. 

> Project Maintainer should have the ability to invite users per https://docs.transifex.com/teams/understanding-user-roles/#teams-&-collaboration

No.
Project Maintainers can only "Assign new project maintainers" per documentation and in real life; or I have not found the correct place: But the one you pointed to does not allow me to do anything (just look at the list) and the one pointed to allows me to add new project maintainers.

Can you also put me in the role "Team Manager" or just invite these three guys to `zh`.

**Comment by nephros** [comment]
> > So this leaves us with the next steps to make the same clean-up in the git repo, plus deleting `src/qml/settings-patchmanager.ts` (and also `translations/settings-patchmanager-en.ts` which is not used?)
> 

I'm not sure about removing the -en.ts one. Formally, the `settings-patchmanager.ts` file can contain any source language (it's only by convention that English is mostly used), and -en.ts should contain proper English translations.  
This is reflected in the fact that `settings-patchmanager.ts` has no language code in its header so even if you build a .qm file from it it (I think) won't be consulted for English l10n.

And from a brief look `settings-patchmanager-en.ts`  actually does contain *some* translations even if all of them are just the same as the source strings.

  
`-en.ts` can also be utilized as I suggested elsewhere, to improve the English strings *without* having to muck about with the source strings - which causes load on the translators and translation commits later which in some cases is unnecessary.

**Comment by Olf0** [comment]
> > > So this leaves us with the next steps to make the same clean-up in the git repo, plus deleting `src/qml/settings-patchmanager.ts` (and also `translations/settings-patchmanager-en.ts` which is not used?)
> 
> I'm not sure about removing the -en.ts one. Formally, the `settings-patchmanager.ts` file can contain any source language (it's only by convention that English is mostly used), and -en.ts should contain proper English translations. This is reflected in the fact that `settings-patchmanager.ts` has no language code in its header so even if you build a .qm file from it it (I think) won't be consulted for English l10n.

Ack, but ...

> And from a brief look `settings-patchmanager-en.ts` actually does contain _some_ translations even if all of them are just the same as the source strings.
> 
> `-en.ts` can also be utilized (as I suggested elsewhere) to improve the English strings _without_ having to muck about with the source strings - which causes load on the translators and translation commits later which in some cases is unnecessary.

Yes, but it is not utilised at Transifex, i.e. English is not a language translated!

And I am addressing this exactly to have this discussion, before taking action: Should it go or not?
IMO it should.

**Comment by b100dian** [comment]
> even if all of them are just the source strings

This sounds like we can add it back if we actually need to not modify the strings for some reason.

While I agree with the fact that this is just an arbitrary language, the practical case is that we are going to have to update it in git manually (just as the source translation is) every time, or we might [even] have surprises.

**Comment by Olf0** [comment]
Done, see PR #108

**Comment by Olf0** [comment]
Because the source file is outdated, I read the documentation on how to update it: https://docs.transifex.com/projects/updating-content/#automatically-updating-source-files
I wonder, if we should enable the automatic update of it?

This time I manually updated it, but what are your preferences?
I am undecided, because "I do not like automatisms, they might turn against you" and the documentation softly warns: If the source file would be accidentially emptied, then that automatically triggers the deletion of all translated strings!  But OTOH, we have a "backup" of the translated files at GH, as long as we do not also automate their download from Transifex.
@b100dian might correctly point out that this extremely unlikely to happen.
And I am undecided (and not softly against it), because I like the idea of the translation process being more decoupled from our work at GH and this is one step less to care / think about.
Likely this decision depends on your opinion, @nephros.

**Comment by Olf0** [comment]
I failed to upload the source TS-file in about 30 tries with FF78ESR and Chromium (which was worse to my surprise).
@b100dian & @nephros: Can one of you please do `curl -LO https://raw.githubusercontent.com/sailfishos-patches/patchmanager/master/translations/settings-patchmanager.ts` and then upload this file under "Resources" with the button "Update source file" at the top to the right.

**Comment by nephros** [comment]
> P.P.S: I am getting this strange warning when uploading that TS-file (when auto-updating no one would see this):
> 
> ```
> English pluralized strings should contain 2 plurals. Instead, on string '.
> ' we found 1
> ```
> 
> What does that mean? Is it relevant?

That comes from https://github.com/sailfishos-patches/patchmanager/blob/master/src/qml/PatchManagerPage.qml#L484 but I'm not sure what it means either. the [docs](https://doc.qt.io/qt-5/i18n-source-translation.html#handling-plurals) say this is how it's done.

**Comment by nephros** [comment]
> I failed to upload the source TS-file in about 30 tries with FF78ESR and Chromium (which was worse to my surprise). @b100dian & @nephros: Can one of you please do `curl -LO https://raw.githubusercontent.com/sailfishos-patches/patchmanager/master/translations/settings-patchmanager.ts` and then upload this file under "Resources" with the button "Update source file" at the top to the right.

Tried it, got the same result. Error message about the plural thing.  
So I manually removed the section containing that string in the .ts file, and *voila!* upload worked.

So please continue with the new updated source strings on Transifex, and I'll try to figure out what's so bad about that source string.

**Comment by nephros** [comment]
> 
> So please continue with the new updated source strings on Transifex, and I'll try to figure out what's so bad about that pluralized source string.

Aha! https://github.com/Sigil-Ebook/Sigil/blob/master/docs/Translating.md

> Now manually edit the resulting base.ts file and change all of the following single numerusform tags:
> to be double empty numerusform tags as follows:
> `<numerusform></numerusform><numerusform></numerusform>`
> so that our base.ts file will work with the much older tools used on the Transifex site.

**Comment by Olf0** [comment]
> > I failed to upload the source TS-file in about 30 tries with FF78ESR and Chromium (which was worse to my surprise). @b100dian & @nephros: Can one of you please do `curl -LO https://raw.githubusercontent.com/sailfishos-patches/patchmanager/master/translations/settings-patchmanager.ts` and then upload this file under "Resources" with the button "Update source file" at the top to the right.
> 
> Tried it, got the same result. Error message about the plural thing. So I manually removed the section containing that string in the .ts file, and _voila!_ upload worked.

Thank you very much.
BTW, my main issue with uploading this file was not that error message: It took quite some additional disabling of NoScript filters (compared to just have the regular functionality of Transifex working) and still it did not work most of the time.  Even when I completely disabled NoScript, rarely the upload succeded (just to see that error message, when it did).  And with Chromium the menu on the right side was not displayed at all, no matter what I tried; usually Chromium works better for such "modern" web-services thickets, as the one Transifex (Atlassian) created here, as Chrome seems to be the primary (or "only") browser web-services programmers test with.

> So please continue with the new updated source strings on Transifex, and I'll try to figure out what's so bad about that pluralized source string.

Thank you for taking care of this issue.
So I will comtinue with what I originally intendend to do, when I started with that two days ago.

> > P.P.S: I am getting this strange warning when uploading that TS-file (when auto-updating no one would see this):
> > ```
> > English pluralized strings should contain 2 plurals. Instead, on string '.
> > ' we found 1
> > ```
> > What does that mean? Is it relevant?
> 
> That comes from https://github.com/sailfishos-patches/patchmanager/blob/master/src/qml/PatchManagerPage.qml#L484 but I'm not sure what it means either. the [docs](https://doc.qt.io/qt-5/i18n-source-translation.html#handling-plurals) say this is how it's done.

Thank you for the analysis.

**Comment by nephros** [comment]
> Thank you for taking care of this issue. So I will comtinue with what I originally intendend to do, when I started with that two days ago.
> 
> > > P.P.S: I am getting this strange warning when uploading that TS-file (when auto-updating no one would see this):
> > > ```
> > > English pluralized strings should contain 2 plurals. Instead, on string '.
> > > ' we found 1
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > What does that mean? Is it relevant?
> > 
> > 
> > That comes from https://github.com/sailfishos-patches/patchmanager/blob/master/src/qml/PatchManagerPage.qml#L484 but I'm not sure what it means either. the [docs](https://doc.qt.io/qt-5/i18n-source-translation.html#handling-plurals) say this is how it's done.
> 
> Thanks for the analysis.

You may want to consider merging this, https://github.com/nephros/patchmanager/tree/l10n-fix, I removed the fancy plural use and separate it into two strings,  so we don't run into that issue again.

**Comment by Olf0** [comment]
> You may want to consider merging this, https://github.com/nephros/patchmanager/tree/l10n-fix, I removed the fancy plural use and separate it into two strings, so we don't run into that issue again.

Even though I wondered, why you did not pose this PR, I went on my way doing this: https://github.com/sailfishos-patches/patchmanager/pull/116
But then I came up with an enhancement: https://github.com/nephros/patchmanager/pull/1/files

So this has become a little complicated:
1. Please merge my PR https://github.com/nephros/patchmanager/pull/1/files
2. Then pose your PR here.
3. As I already have reviewed your PR, I will mark that as reviewed as soon I can.
4. If you want me to merge it right away, please denote that.

**Comment by Olf0** [comment]
@nephros, thank you for uploading the reworked TS source file.  (I tried that, failed again, then looked at the timestamps, only to realise that you have been quite quick in uploading it after I merged #117.)

Thant brings us back to my original question (which was the start of a lengthy detour): [I wonder, if we should enable the automatic update of it?](https://github.com/sailfishos-patches/patchmanager/issues/27#issuecomment-955636859)

Due to my repeated issues with uploading the TS source file, I now unambiguously vote for "Yes"!
If another "Yes" vote is cast, I will happily employ this change at Transifex's settings.

**Comment by Olf0** [comment]
Do note the excurse with documentation and decisions at https://github.com/sailfishos-patches/patchmanager/pull/145#issuecomment-960222739 ff.!

**Comment by Olf0** [comment]
Oh, I was about to close this issue, but one thing is left to do:

Document [the decision to accept only primary languages](https://github.com/sailfishos-patches/patchmanager/issues/27#issuecomment-937402501) in [Patchmanager's Wiki section for Translations ("i18n", "l10n")](https://github.com/sailfishos-patches/patchmanager/wiki/Translations-(%22i18n%22,-%22l10n%22)).
